pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,356 @@
+// sync.Mutex: is required to handle race conditions for
+// storing and retrieving tasks during a driver restart
+// taskMap: is an in-memory map to associate task details
+// to the channel created by the caller to receive the task result
+// context.Context: is separate from the context set by CSI ops
+// listView: holds the managed object used to monitor multiple concurrent VC tasks
+// shouldStopListening: in case of regular CSI operation, even after receiving a batch of updates,
+// we want to continue listening for subsequent updates.
+// in case of unit tests, we need to stop listening for the test to complete execution
+type TaskViewManagerImpl struct {
+	mu                  sync.Mutex
+	virtualCenter       *cnsvsphere.VirtualCenter
+	listView            *view.ListView
+	ctx                 context.Context
+	taskMap             map[types.ManagedObjectReference]*TaskDetails
+	shouldStopListening bool
+}
+
+func (t *TaskViewManagerImpl) SetVirtualCenter(virtualCenter *cnsvsphere.VirtualCenter) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.virtualCenter = virtualCenter
+}
+
+func getListViewWaitFilter(listView *view.ListView) *property.WaitFilter {
+	ts := types.TraversalSpec{
+		Type: "ListView",
+		Path: "view",
+		Skip: types.NewBool(false),
+	}
+	filter := new(property.WaitFilter)
+
+	filter.Add(listView.Reference(), "Task", []string{infoPropertyName}, &ts)
+	reportMissingObjectsInResults := true
+	filter.Spec.ReportMissingObjectsInResults = &reportMissingObjectsInResults
+	return filter
+}
+
+// listenToTaskUpdates is a long-running goroutine
+// that uses a property collector to listen for task updates
+// CSI ops add CNS tasks to listview and wait for a response from CNS
+// when update(s) are received by the property collector,
+// it spawns a new goroutine to process each task update and return the result to the caller
+func (t *TaskViewManagerImpl) listenToTaskUpdates(ctx context.Context) {
+	log := logger.GetLogger(ctx)
+	var filter *property.WaitFilter
+	recreateView := false
+	for {
+		if t.getListView() == nil || recreateView {
+			log.Debug("re-creating the listView object")
+			listView, err := createTasksListView(ctx, t.virtualCenter.Client.Client, nil)
+			if err != nil {
+				log.Errorf("failed to create a ListView. error: %+v", err)
+				time.Sleep(waitForUpdatesRetry)
+				continue
+			}
+			t.setListView(t.ctx, listView)
+		}
+
+		if filter == nil || recreateView {
+			filter = getListViewWaitFilter(t.getListView())
+			recreateView = false
+		}
+		log.Debug("Starting listening for updates...")
+		pc := property.DefaultCollector(t.virtualCenter.Client.Client)
+		err := property.WaitForUpdates(t.ctx, pc, filter, func(updates []types.ObjectUpdate) bool {
+			log.Debugf("Got %d property collector update(s)", len(updates))
+			for _, update := range updates {
+				for _, prop := range update.ChangeSet {
+					log.Debugf("Got update for object %v properties %v", update.Obj, prop)
+					if prop.Name == infoPropertyName {
```

Comment:
unless we anticipate updates from multiple properties, we do not need to check the name.

Reply:
Fixed.